feat: add support for HALF_VECTOR column type

yberber-sap opened the pull request:
- Added a `VectorColumnType` enum to represent vector column types.
- Added a new `vector_column_type` parameter in the `__init__` and `from_texts` methods with a default of `REAL_VECTOR`.
- Modified `_initialize_table` to use the specified vector column type when creating the table. If the table has already been created but the vector column type is not correctly given, it will fail.
- Updated `_serialize_fvecs` and `_deserialize_fvecs` to handle both `REAL_VECTOR` (4-byte) and `HALF_VECTOR` (2-byte) serialization.
- Added `_get_vector_conversion_function` to dynamically return the correct vector conversion function (TO_REAL_VECTOR or TO_HALF_VECTOR).
- Modified `_add_texts_using_external_embedding` and `_add_texts_using_internal_embedding` to support both vector types.
- Updated `similarity_search_with_score_and_vector_by_vector` to use the appropriate vector conversion function.
- Introduced `_conditionally_wrap_half_vector` method to wrap expressions with TO_HALF_VECTOR(...) when the vector column type is HALF_VECTOR. This is used to wrap `VECTOR_EMBEDDING(....)` statement which returns `REAL_VECTOR`

I was unable to fully test the case where both internal embedding and HALF_VECTOR are used together, as the instance I tested `HALF_VECTOR` on does not support the VECTOR_EMBEDDING function. Based on the presentation, I assumed that V`ECTOR_EMBEDDING` returns `REAL_VECTOR`, and therefore needs to be wrapped with `TO_HALF_VECTOR(...)  when storing into HALF_VECTOR column or performing similarity search with embedded user query and vector column values. This is done using `_conditionally_wrap_half_vector`

fredi-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -1055,7 +1130,7 @@ def max_marginal_relevance_search(  # type: ignore[override]
-                    embedding = HanaDB._deserialize_fvecs(res[0][0])
+                    embedding = self._deserialize_fvecs(res[0][0])
             finally:
                 cur.close()
```
Since the `if` block is a one-liner, it might be nice to create a small helper function for the `else` block in order to have (more or less) a single level of abstraction.

yberber-sap replied:
Ah, I see—great point! I actually considered extracting the `else` block into a separate method as well. The main reason I hesitated is because we don’t want users to directly use the `VECTOR_EMBEDDING` function to embed their queries, unlike how external embedding instances are typically used.

For example, if someone tries:

```python
internal_embedding = HanaInternalEmbeddings("SAP_NEB.20240715")
internal_embedding.embed_query("Hello")
```

It would raise an error, since that’s not the intended usage.

Calling something like `self._embed_query_hana_internal()` would serve as an alternative to `internal_embedding.embed_query(user_query)` 

Also, in methods like `similarity_search` or `add_texts`, we don’t retrieve the embedding of the user query. But for `max_marginal_relevance_search`, it's essential.

I will name it as `_embed_query_hana_internal()` (with _), to clarify that it’s for internal use.

yberber-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -62,6 +62,12 @@
 
 
+# Enum to represent vector column types
+class VectorColumnType:
```
I initially used an `Enum`, but I found it a bit inconvenient from the user's perspective depending on where the enum is defined:

- **If defined as an inner class** inside `HanaDB`, users have to access it like this:
  ```python
  HanaDB(..., vector_column_type=HanaDB.VectorColumnType.REAL_VECTOR)
  ```
- **If defined at the module level** (outside the `HanaDB` class), then they need to import it explicitly:
  ```python
  from vectorstores import VectorColumnType
  HanaDB(..., vector_column_type=VectorColumnType.REAL_VECTOR)
  ```

I don't know if there is a better implementation with `Enum`

Additionally, even when using an `Enum`, **type checking can still be necessary** at runtime to guard against invalid inputs. For example, a user can still pass anything If I am not wrong:

```python
if not isinstance(vector_column_type, VectorColumnType):
    raise TypeError(f"Expected VectorColumnType, got {type(vector_column_type).__name__}")
```

I also didn’t go with a plain list of strings like `["REAL_VECTOR", "HALF_VECTOR"]`, because I wanted to avoid hardcoded string literals throughout the codebase.

That said, hardcoding isn’t a huge issue in this specific case, since the corresponding column types in HANA are unlikely to change.

Based on your feedback, I’m open to adjusting the implementation—either by using a properly validated `Enum` or simplifying it with a constant list like:

```python
VECTOR_COLUMN_SQL_TYPES = ["REAL_VECTOR", "HALF_VECTOR"]
```

Let me know what you think.

yberber-sap replied:
I used the following for vector column type handling:
```
VECTOR_COLUM_SQL_TYPES = ["REAL_VECTOR", "HALF_VECTOR"]
```

yberber-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -282,15 +292,36 @@ def _sanitize_specific_metadata_columns(
+            )
+        return vector_column_type
+
+    def _serialize_fvecs(self, values: list[float]) -> bytes:
```
I renamed to `_serialize_binary_format` and `_deserialize_binary_format` and adapted the comments

yberber-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -282,15 +292,36 @@ def _sanitize_specific_metadata_columns(
+        return vector_column_type
+
+    def _serialize_fvecs(self, values: list[float]) -> bytes:
         # Converts a list of floats into FVECS binary format
```
I adapted it as follows
```    
    def _serialize_binary_format(self, values: list[float]) -> bytes:
        # Converts a list of floats into binary format
        if self.vector_column_type == VectorColumnType.HALF_VECTOR:
            # 2-byte half-precision float serialization
            return struct.pack(f"<I{len(values)}e", len(values), *values)
        else:
            # 4-byte float serialization (standard FVECS format)
            return struct.pack(f"<I{len(values)}f", len(values), *values)```

yberber-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -282,15 +292,36 @@ def _sanitize_specific_metadata_columns(
-        return list(struct.unpack_from("<%sf" % dim, fvecs, 4))
+        if self.vector_column_type == VectorColumnType.HALF_VECTOR:
+            # 2-byte half-precision float deserialization
+            return list(struct.unpack_from("<%se" % dim, fvecs, 4))
```
Thanks, good point! I updated both format strings to use f-strings

yberber-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -62,6 +62,12 @@
 default_vector_column_length: int = -1  # -1 means dynamic length
```
you're right. I renamed

yberber-sap replied:
But for other column names, we didn't used the pattern with the prefix `_name`. Can we ignore the inconsistency? 
```
default_content_column: str = "VEC_TEXT"
default_metadata_column: str = "VEC_META"
default_vector_column_name: str = "VEC_VECTOR"

class HanaDB(VectorStore):
    def __init__(...):
        self.content_column = HanaDB._sanitize_name(content_column)
        self.metadata_column = HanaDB._sanitize_name(metadata_column)
        self.vector_column_name = HanaDB._sanitize_name(vector_column_name)
```

yberber-sap replied:
Then I leave as it is for now:)

yberber-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -304,6 +335,36 @@ def _split_off_special_metadata(self, metadata: dict) -> tuple[dict, list]:
+        elif self.vector_column_type == VectorColumnType.HALF_VECTOR:
+            return "TO_HALF_VECTOR"
+        else:
+            raise ValueError(f"Unsupported vector type: {self.vector_column_type}")
```
I adapted the method _get_vector_conversion_function(expr) as follows:
```
    def _convert_to_target_vector_type(self, expr: str) -> str:
        if self.vector_column_type in VECTOR_COLUMN_SQL_TYPES:
            return f"TO_{self.vector_column_type}('{expr}')"
        else:
            raise ValueError(f"Unsupported vector type: {self.vector_column_type}")
```
   

This new version wraps the expression with the corresponding vector column type as suggested in:

```
    def _convert_vector_embedding_to_column_type(self, expr: str) -> str:
        """
        Makes sure that an embedding produced by HANA's VECTOR_EMBEDDING
        aligns with the column type.

        Note that VECTOR_EMBEDDING always returns REAL_VECTORs.

        Args:
            expr (str): SQL expression producing an embedding vector.
        Returns:
            str: Wrapped expression if vector column type is not REAL_VECTOR,
                otherwise the original expression.
        """

        if "VECTOR_EMBEDDING" not in expr.upper():
            raise ValueError(f"Expected 'VECTOR_EMBEDDING' in '{expr}'")

        if self.vector_column_type != "REAL_VECTOR":
            return f"{self._get_vector_conversion_function()}({expr})"
            # consider passing an expression to conversion function
            # return self._get_vector_conversion_function(expr)      <<<<--------

        return expr
```

yberber-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -282,15 +292,36 @@ def _sanitize_specific_metadata_columns(
+        if self.vector_column_type == VectorColumnType.HALF_VECTOR:
+            # 2-byte half-precision float serialization
+            return struct.pack(f"<I{len(values)}e", len(values), *values)
+        else:
```
thanks! done

yberber-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -559,6 +626,7 @@ def from_texts(  # type: ignore[no-untyped-def, override]
             metadata_column=metadata_column,
             vector_column=vector_column,
             vector_column_length=vector_column_length,  # -1 means dynamic length
```
Thanks! In a separate PR, I will deal with it :)

fredi-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -292,36 +289,40 @@ def _sanitize_specific_metadata_columns(
     def _serialize_binary_format(self, values: list[float]) -> bytes:
         # Converts a list of floats into binary format
-        if self.vector_column_type == VectorColumnType.HALF_VECTOR:
+        if self.vector_column_type == "HALF_VECTOR":
```
One could argue that sql types are case insensitive, so you should do a case insensitive comparison. But this is something for a follow-up.

yberber-sap replied:
Thanks for the tip! I thought they were case-sensitive.  
I’ve updated the `_sanitize_vector_column_type(vector_column_type: str)` method to return the vector column type in uppercase, which is then used to assign `self.vector_column_type` in the `__init__` method:

```python
self.vector_column_type = HanaDB._sanitize_vector_column_type(
    vector_column_type
)
```


_sanitize_vector_column_type:
```
    def _sanitize_vector_column_type(vector_column_type: str) -> str:
        vector_column_type_upper = vector_column_type.upper()    <<<<----
        if vector_column_type_upper not in VECTOR_COLUMN_SQL_TYPES:
            raise ValueError(
                f"Unsupported vector_column_type: {vector_column_type}. "
                f"Must be one of {', '.join(VECTOR_COLUMN_SQL_TYPES)}"
            )
        return vector_column_type_upper.    <<<----
```

fredi-sap replied:
Nice! I've got nothing more to add. 👍

yberber-sap commented:
> First small review comments. Additionally to the comments in the code:
> 
> 1. Can we add tests to the new functionalities?

Of course! But right now, I have no instance supporting HALF_VECTOR! I will add the tests ASAP! Last week, I changed the default vector column type parameter from REAL_VECTOR to HALF_VECTOR. Everything was working pretty good!

> 2. I think we need to check if the HANA instance we are connected to does support HALF_VECTOR (via Version)

I added a check for HALF_VECTOR. Let me quickly explain what I did, since the PR touches multiple lines and this particular change might not be immediately obvious.

We only call HanaDB._sanitize_vector_column_type once, inside the __init__. I modified this function to additionally call HanaDB._check_half_vector_availability when the passed vector_column_type is "HALF_VECTOR".

The _check_half_vector_availability method runs the SQL command SELECT TYPE_NAME FROM SYS.DATA_TYPES to retrieve the list of available datatypes in the current HANA instance. If "HALF_VECTOR" is not present in that list, we raise an error.

Right now, I’m using a ValueError with the message:

"HALF_VECTOR is only supported on instances from 2025.15 (QRC 2/2025) onward."

I'm not entirely sure if ValueError is the best choice here—it works, since the issue is with an invalid parameter in this environment, and we’ve used it similarly elsewhere. However, NotImplementedError might also make sense, given that the feature is version-dependent and not yet supported in some environments.

Here’s the original function:

```
@staticmethod
def _sanitize_vector_column_type(vector_column_type: str) -> str:
    vector_column_type_upper = vector_column_type.upper()
    if vector_column_type_upper not in VECTOR_COLUMN_SQL_TYPES:
        raise ValueError(
            f"Unsupported vector_column_type: {vector_column_type}. "
            f"Must be one of {', '.join(VECTOR_COLUMN_SQL_TYPES)}"
        )
    return vector_column_type_upper
```

And here’s the updated version with the check:

```
@staticmethod
def _sanitize_vector_column_type(
    vector_column_type: str, connection: dbapi.Connection
) -> str:
    vector_column_type_upper = vector_column_type.upper()
    if vector_column_type_upper not in VECTOR_COLUMN_SQL_TYPES:
        raise ValueError(
            f"Unsupported vector_column_type: {vector_column_type}. "
            f"Must be one of {', '.join(VECTOR_COLUMN_SQL_TYPES)}"
        )
    if vector_column_type_upper == "HALF_VECTOR":
        HanaDB._check_half_vector_availability(connection)
    return vector_column_type_upper

@staticmethod
def _check_half_vector_availability(connection: dbapi.Connection):
    cur = connection.cursor()
    try:
        cur.execute("SELECT TYPE_NAME FROM SYS.DATA_TYPES")
        if cur.has_result_set():
            rows = cur.fetchall()
            available_types = {row[0] for row in rows}
            if "HALF_VECTOR" in available_types:
                return True
        raise ValueError(
            "HALF_VECTOR is only supported on instances from 2025.15 (QRC 2/2025) onward."
        )
    finally:
        cur.close()
```

Let me know if you'd prefer switching to NotImplementedError or defining a custom exception for clarity.

fredi-sap commented on langchain_hana/vectorstores.py:
```diff
@@ -288,16 +288,37 @@ def _sanitize_specific_metadata_columns(
                 f"Must be one of {', '.join(VECTOR_COLUMN_SQL_TYPES)}"
             )
+        if vector_column_type_upper == "HALF_VECTOR":
+            HanaDB._check_half_vector_availability(connection)
```
I would continue as in line 297:

```python3
        if vector_column_type_upper not in self._getAvailableHanaDataTypes():
            raise ValueError(
                f"Your HANA installation does not have {vector_column_type=}. "
                f"The HANA version is {self._getHanaVersion()}"
            )
```

A `ValueError` is fine. I would only use `NotImplementedError` for stuff which is not implemented in your plugin.

Note the snippet above also checks if your HANA installation supports `REAL_VECTOR`. You never know...

fredi-sap replied:
Rereading the code, you might raise a `NotImplementedError` in line 298. But I would keep the `ValueError` for the HANA datatype check.

yberber-sap replied:
> I would continue as in line 297:
> 
> ```python
>         if vector_column_type_upper not in self._getAvailableHanaDataTypes():
>             raise ValueError(
>                 f"Your HANA installation does not have {vector_column_type=}. "
>                 f"The HANA version is {self._getHanaVersion()}"
>             )
> ```
> 
> A `ValueError` is fine. I would only use `NotImplementedError` for stuff which is not implemented in your plugin.
> 
> Note the snippet above also checks if your HANA installation supports `REAL_VECTOR`. You never know...

Thanks! You're right—I'll stick with raising a `ValueError` and will also add the check for the `REAL_VECTOR` datatype.

As I understand it, `self._getHanaVersion()` returns the version of the currently running HANA instance. Would it maybe be more helpful to instead return the **minimum required version** that supports the given datatype? For example:

```
        if vector_column_type_upper not in self._getAvailableHanaDataTypes():
            raise ValueError(
                f"{vector_column_type_upper} is only supported on instances from {self.get_min_version(vector_column_type_upper)} onward."
            )
```


> Rereading the code, you might raise a `NotImplementedError` in line 298. But I would keep the `ValueError` for the HANA datatype check.

If the user provides something like `"HELLO_WORLD"` as `vector_column_type`, raising a `NotImplementedError` could work—but I think `ValueError` feels a little bit better here. Because it's not about a feature we'll implement later, just invalid value

fredi-sap replied:
> As I understand it, `self._getHanaVersion()` returns the version of the currently running HANA instance.

Yes, that's what the function should do.

> Would it maybe be more helpful to instead return the minimum required version that supports the given datatype?

Sure. Consider printing both the installed HANA version and the minimum required version.

> If the user provides something like ... `ValueError` feels a little bit better here ...

OK, got you, then leave it as `ValueError`.
